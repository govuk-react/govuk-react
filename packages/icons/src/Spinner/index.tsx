// https://govuk-loader-prototype.herokuapp.com/components/loader

import React from 'react';
import styled, { keyframes } from 'styled-components';

import SVG, { SVGProps } from '../SVGBase/index';

const fadeInOut = keyframes`
  0% { opacity: 0.250075; }
  0.01% { opacity: 0.25; }
  0.03% { opacity: 1; }
  100% { opacity: 0.250075; }
`;

const Rect = styled.rect`
  animation: ${fadeInOut} 1s infinite linear;
`;

interface SpinnerProps extends SVGProps {
  className?: string;
  title?: string;
  fill?: string;
}

const Spinner: React.FC<SpinnerProps> = ({ className, fill, title, ...rest }: SpinnerProps) => (
  <SVG
    className={className}
    viewBox="-25 -25 50 50"
    preserveAspectRatio="xMidYMid meet"
    title={title}
    fill={fill}
    {...rest}
  >
    {Array(12)
      .fill(1)
      .map((el, i) => (
        <Rect
          fill={fill}
          width="12"
          height="5"
          rx="2.5"
          ry="2.5"
          style={{ animationDelay: `${i * 83}ms` }}
          transform={`rotate(${i * 30}, 0, 2) translate(10 0)`}
          opacity="0"
          /* eslint-disable-next-line react/no-array-index-key */
          key={i}
        />
      ))}
  </SVG>
);

Spinner.defaultProps = {
  className: 'icon-spinner',
  title: 'Loading',
  fill: 'currentColor',
};

export default Spinner;
