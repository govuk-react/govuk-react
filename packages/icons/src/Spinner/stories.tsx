import React from 'react';

import { withKnobs, color, text } from '@storybook/addon-knobs';

import SpinnerComponent from '.';

export default {
  title: 'Utility/Icons',
  decorators: [withKnobs],
};

export const Spinner: React.FC = () => {
  const fillColour = color('fill', 'black');
  const width = text('width', '256px');
  const height = text('height', '256px');
  return <SpinnerComponent width={width} height={height} fill={fillColour} title="Example Search implementation" />;
};
