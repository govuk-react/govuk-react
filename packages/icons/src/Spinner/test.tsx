import * as React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import Spinner from '.';

const wrapper = (
  <Spinner width="100px" height="100px" fill="red">
    Example
  </Spinner>
);

describe('Spinner', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(wrapper, div);
  });

  it('matches wrapper snapshot', () => {
    expect(mount(wrapper)).toMatchSnapshot('wrapper mount');
  });
});
