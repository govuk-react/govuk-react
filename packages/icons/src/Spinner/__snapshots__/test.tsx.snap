// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner matches wrapper snapshot: wrapper mount 1`] = `
.c0 {
  display: block;
}

.c1 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 0ms;
  animation-delay: 0ms;
}

.c2 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 83ms;
  animation-delay: 83ms;
}

.c3 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 166ms;
  animation-delay: 166ms;
}

.c4 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 249ms;
  animation-delay: 249ms;
}

.c5 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 332ms;
  animation-delay: 332ms;
}

.c6 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 415ms;
  animation-delay: 415ms;
}

.c7 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 498ms;
  animation-delay: 498ms;
}

.c8 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 581ms;
  animation-delay: 581ms;
}

.c9 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 664ms;
  animation-delay: 664ms;
}

.c10 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 747ms;
  animation-delay: 747ms;
}

.c11 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 830ms;
  animation-delay: 830ms;
}

.c12 {
  -webkit-animation: hhGhiE 1s infinite linear;
  animation: hhGhiE 1s infinite linear;
  -webkit-animation-delay: 913ms;
  animation-delay: 913ms;
}

<Spinner
  className="icon-spinner"
  fill="red"
  height="100px"
  title="Loading"
  width="100px"
>
  <SVG
    className="icon-spinner"
    fill="red"
    height="100px"
    preserveAspectRatio="xMidYMid meet"
    title="Loading"
    viewBox="-25 -25 50 50"
    width="100px"
  >
    <SVGBase__StyledSvg
      className="icon-spinner"
      fill="red"
      height="100px"
      preserveAspectRatio="xMidYMid meet"
      version="1.1"
      viewBox="-25 -25 50 50"
      width="100px"
    >
      <svg
        className="c0 icon-spinner"
        fill="red"
        height="100px"
        preserveAspectRatio="xMidYMid meet"
        version="1.1"
        viewBox="-25 -25 50 50"
        width="100px"
      >
        <title>
          Loading
        </title>
        <Spinner__Rect
          animationDelay={0}
          fill="red"
          height="5"
          key="0"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(0, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c1"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(0, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={1}
          fill="red"
          height="5"
          key="1"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(30, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c2"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(30, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={2}
          fill="red"
          height="5"
          key="2"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(60, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c3"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(60, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={3}
          fill="red"
          height="5"
          key="3"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(90, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c4"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(90, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={4}
          fill="red"
          height="5"
          key="4"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(120, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c5"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(120, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={5}
          fill="red"
          height="5"
          key="5"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(150, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c6"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(150, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={6}
          fill="red"
          height="5"
          key="6"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(180, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c7"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(180, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={7}
          fill="red"
          height="5"
          key="7"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(210, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c8"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(210, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={8}
          fill="red"
          height="5"
          key="8"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(240, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c9"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(240, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={9}
          fill="red"
          height="5"
          key="9"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(270, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c10"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(270, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={10}
          fill="red"
          height="5"
          key="10"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(300, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c11"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(300, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
        <Spinner__Rect
          animationDelay={11}
          fill="red"
          height="5"
          key="11"
          opacity="0"
          rx="2.5"
          ry="2.5"
          transform="rotate(330, 0, 2) translate(10 0)"
          width="12"
        >
          <rect
            className="c12"
            fill="red"
            height="5"
            opacity="0"
            rx="2.5"
            ry="2.5"
            transform="rotate(330, 0, 2) translate(10 0)"
            width="12"
          />
        </Spinner__Rect>
      </svg>
    </SVGBase__StyledSvg>
  </SVG>
</Spinner>
`;
